plugin/src/com/microsoft/alm/plugin/idea/tfvc/core/TfvcIntegrationEnabler.java

```diff
@@ -0,0 +1,304 @@
+import org.jetbrains.annotations.NotNull;
+import org.jetbrains.annotations.Nullable;
+
+import java.nio.file.Path;
+import java.nio.file.Paths;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+import java.util.List;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.CompletionStage;
+
+import static java.util.stream.Collectors.toList;
+
+public class TfvcIntegrationEnabler extends VcsIntegrationEnabler {
+
+    private static final Logger ourLogger = Logger.getInstance(TfvcIntegrationEnabler.class);
+
+    private final TFSVcs myVcs;
+
+    protected TfvcIntegrationEnabler(@NotNull TFSVcs vcs) {
+        super(vcs);
+        myVcs = vcs;
+    }
+
+    @Override
+    public void enable(@NotNull Collection<VcsRoot> vcsRoots) {
+        // This override does the same as base method, but tries to determine a workspace directory instead of using
+        // project.getBaseDir().
+        Collection<VirtualFile> existingRoots = vcsRoots.stream().filter(root -> {
+            AbstractVcs<?> vcs = root.getVcs();
+            return vcs != null && vcs.getName().equals(myVcs.getName());
+        }).map(VcsRoot::getPath).collect(toList());
+
+        if (!existingRoots.isEmpty()) {
+            super.enable(vcsRoots);
+            return;
+        }
+
+        String basePath = myProject.getBasePath();
+        if (basePath == null) {
+            ourLogger.warn("Project base path is null");
+            return;
+        }
+
+        Path workspacePath = determineWorkspaceDirectory(Paths.get(basePath));
+        VirtualFile workspaceFile = ObjectUtils.notNull(
+                LocalFileSystem.getInstance().findFileByIoFile(workspacePath.toFile()));
+
+        if (initOrNotifyError(workspaceFile))
+            addVcsRoot(workspaceFile);
+    }
+
+    @NotNull
+    private static Path determineWorkspaceDirectory(@NotNull Path projectBasePath) {
+        Path vsClient = VisualStudioTfvcClient.getOrDetectPath(PropertyService.getInstance());
+        Path path = projectBasePath;
+        do {
+            Workspace workspace = null;
+            try {
+                ourLogger.info("Analyzing path \"" + path + "\" using TF Everywhere client");
+                workspace = CommandUtils.getPartialWorkspace(path);
+            } catch (WorkspaceCouldNotBeDeterminedException ex) {
+                ourLogger.info("Path \"" + path + "\" has no TF Everywhere workspace");
+            }
+
+            if (workspace == null && vsClient != null)
+                workspace = VisualStudioTfvcCommands.getPartialWorkspaceAsync(vsClient, path)
+                        .toCompletableFuture().join();
+
+            String currentPath = path.toAbsolutePath().toString();
+            boolean correspondsToAnyMapping = workspace != null && workspace.getMappings().stream()
+                    .anyMatch(mapping -> FileUtil.pathsEqual(currentPath, mapping.getLocalPath()));
```

Comment:
Could we have any problem here? We compare string. Maybe we should normalize separators and short long paths.

Reply:
Or convert to system independent paths.

Reply:
`FileUtil.pathsEqual` already peforms the necessary normalization, and even tries to handle case-insensitive file systems somewhat properly. I believe this is the best approach we should take here.